File: backend/routers/case.py
from fastapi import APIRouter, Depends, Query, HTTPException
from sqlalchemy.orm import Session
from sqlalchemy import func
from typing import List, Optional
import os
from schemas.case import CaseItem, CaseDetail, CaseUpdate, CaseCreate
from schemas.base import BaseResponse
from config.database import get_db
from models.case import Case
from models.material import Material
from models.chat import ChatMessage
from models.user import User
from utils.deps import get_current_user

router = APIRouter(prefix="/cases", tags=["案件管理"])

@router.post("", response_model=BaseResponse[CaseDetail])
async def create_case(
    case_data: CaseCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    new_case = Case(
        user_id=current_user.id,
        title=case_data.title,
        parties=case_data.parties,
        phone=case_data.phone,
        case_type=case_data.case_type,
        status=case_data.status,
        amount=case_data.amount,
        description=case_data.description,
        material_count=0,
        total_file_size="0B"
    )
    
    db.add(new_case)
    db.commit()
    db.refresh(new_case)
    
    new_case.material_stats = {
        "case": 0,
        "evidence": 0,
        "payment": 0,
        "notice": 0
    }
    return BaseResponse.success(result=CaseDetail.from_orm(new_case))

@router.get("", response_model=BaseResponse[List[CaseItem]])
async def get_cases(
    search: Optional[str] = Query(None, description="搜索关键词"),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    query = db.query(Case).filter(Case.user_id == current_user.id)
    
    if search:
        query = query.filter(
            (Case.parties.ilike(f"%{search}%")) | 
            (Case.case_type.ilike(f"%{search}%")) |
            (Case.title.ilike(f"%{search}%"))
        )
        
    cases = query.all()
    
    # 为每个案件计算材料数量和文件总大小
    for c in cases:
        materials = db.query(Material).filter(Material.case_id == c.id).all()
        c.material_count = len(materials)
        
        total_bytes = 0
        for mat in materials:
            size_str = mat.file_size
            if size_str.endswith('KB'):
                total_bytes += float(size_str[:-2]) * 1024
            elif size_str.endswith('MB'):
                total_bytes += float(size_str[:-2]) * 1024 * 1024
            elif size_str.endswith('GB'):
                total_bytes += float(size_str[:-2]) * 1024 * 1024 * 1024
            elif size_str.endswith('B'):
                total_bytes += float(size_str[:-1])
        
        if total_bytes >= 1024 * 1024 * 1024:
            c.total_file_size = f"{total_bytes / (1024 * 1024 * 1024):.2f}GB"
        elif total_bytes >= 1024 * 1024:
            c.total_file_size = f"{total_bytes / (1024 * 1024):.2f}MB"
        elif total_bytes >= 1024:
            c.total_file_size = f"{total_bytes / 1024:.2f}KB"
        else:
            c.total_file_size = f"{total_bytes}B"
        
        c.created_at = c.created_at.strftime("%Y-%m-%d %H:%M")
        
    return BaseResponse.success(result=cases)

@router.get("/{case_id}", response_model=BaseResponse[CaseDetail])
async def get_case_detail(
    case_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    case_obj = db.query(Case).filter(Case.id == case_id, Case.user_id == current_user.id).first()
    if not case_obj:
        return BaseResponse.error(msg="案件不存在")
    
    material_stats = {
        "case": 0,
        "evidence": 0,
        "payment": 0,
        "notice": 0
    }

    materials = db.query(Material).filter(Material.case_id == case_id).all()
    for mat in materials:
        if mat.category in material_stats:
            material_stats[mat.category] += 1

    case_obj.material_stats = material_stats

    return BaseResponse.success(result=CaseDetail.from_orm(case_obj))

@router.put("/{case_id}", response_model=BaseResponse[CaseDetail])
async def update_case(
    case_id: int,
    case_data: CaseUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    case_obj = db.query(Case).filter(Case.id == case_id, Case.user_id == current_user.id).first()
    if not case_obj:
        raise HTTPException(status_code=404, detail="案件不存在")
    
    update_data = case_data.model_dump(exclude_unset=True)
    for key, value in update_data.items():
        setattr(case_obj, key, value)
    
    db.commit()
    db.refresh(case_obj)
    
    material_stats = {
        "case": 0,
        "evidence": 0,
        "payment": 0,
        "notice": 0
    }

    materials = db.query(Material).filter(Material.case_id == case_id).all()
    for mat in materials:
        if mat.category in material_stats:
            material_stats[mat.category] += 1

    case_obj.material_stats = material_stats
    
    return BaseResponse.success(result=CaseDetail.from_orm(case_obj))

@router.delete("/{case_id}", response_model=BaseResponse[None])
async def delete_case(
    case_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    case_obj = db.query(Case).filter(Case.id == case_id, Case.user_id == current_user.id).first()
    if not case_obj:
        raise HTTPException(status_code=404, detail="案件不存在")
    
    materials = db.query(Material).filter(Material.case_id == case_id).all()
    
    BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    for mat in materials:
        if mat.file_url:
            file_path = os.path.join(BASE_DIR, mat.file_url.lstrip("/"))
            if os.path.exists(file_path):
                try:
                    os.remove(file_path)
                except:
                    pass
    
    db.query(Material).filter(Material.case_id == case_id).delete()
    
    db.query(ChatMessage).filter(ChatMessage.case_id == case_id).delete()
    
    db.delete(case_obj)
    db.commit()
    
    return BaseResponse.success(result=None, msg="删除成功")

File: frontend/lib/viewmodels/case.dart
import 'package:flutter/material.dart';

class Case {
  String id;
  String user_id;
  String title;
  String case_type;
  String status;
  String create_at;
  String parties;
  String material_count;
  String total_file_size;
  String phone;

  Case({
    required this.id,
    required this.user_id,
    required this.title,
    required this.case_type,
    required this.status,
    required this.create_at,
    required this.parties,
    required this.material_count,
    required this.total_file_size,
    required this.phone,
  });

  factory Case.fromJSON(Map<String, dynamic> json) {
    return Case(
      id: json["id"]?.toString() ?? "",
      user_id: json["user_id"]?.toString() ?? "",
      title: json["title"]?.toString() ?? "",
      case_type: json["case_type"]?.toString() ?? "",
      status: json["status"]?.toString() ?? "",
      create_at: json["created_at"]?.toString() ?? "", // 修改为 created_at
      parties: json["parties"]?.toString() ?? "",
      material_count: json["material_count"]?.toString() ?? "",
      total_file_size: json["total_file_size"]?.toString() ?? "",
      phone: json["phone"]?.toString() ?? "",
    );
  }
}

class CaseDetail {
  String id;
  String user_id;
  String title;
  String parties;
  String phone;
  String case_type;
  String status;
  String amount;
  String description;
  Map<String, dynamic> material_stats;

  CaseDetail({
    required this.id,
    required this.user_id,
    required this.title,
    required this.parties,
    required this.phone,
    required this.case_type,
    required this.status,
    required this.amount,
    required this.description,
    required this.material_stats,
  });

  factory CaseDetail.fromJSON(Map<String, dynamic> json) {
    return CaseDetail(
      id: json["id"]?.toString() ?? "",
      user_id: json["user_id"]?.toString() ?? "",
      title: json["title"]?.toString() ?? "",
      parties: json["parties"]?.toString() ?? "",
      phone: json["phone"]?.toString() ?? "",
      case_type: json["case_type"]?.toString() ?? "",
      status: json["status"]?.toString() ?? "",
      amount: json["amount"]?.toString() ?? "0",
      description: json["description"]?.toString() ?? "",
      material_stats: json["material_stats"] ?? {},
    );
  }
}

class Material {
  String id;
  String case_id;
  String user_id;
  String category; // case, evidence, payment, notice
  String name;
  String? file_type;
  String file_size;
  String? file_url;
  String? content;
  String created_at;
  String? updated_at;

  Material({
    required this.id,
    required this.case_id,
    required this.user_id,
    required this.category,
    required this.name,
    this.file_type,
    required this.file_size,
    this.file_url,
    this.content,
    required this.created_at,
    this.updated_at,
  });

  factory Material.fromJSON(Map<String, dynamic> json) {
    return Material(
      id: json["id"]?.toString() ?? "",
      case_id: json["case_id"]?.toString() ?? "",
      user_id: json["user_id"]?.toString() ?? "",
      category: json["category"]?.toString() ?? "",
      name: json["name"]?.toString() ?? "",
      file_type: json["file_type"]?.toString(),
      file_size: json["file_size"]?.toString() ?? "0B",
      file_url: json["file_url"]?.toString(),
      content: json["content"]?.toString(),
      created_at: json["created_at"]?.toString() ?? "",
      updated_at: json["updated_at"]?.toString(),
    );
  }
}
